crates/core/tedge_core/src/core_task.rs

```diff
@@ -0,0 +1,50 @@
+use std::collections::HashMap;
+
+use tedge_api::address::EndpointKind;
+
+use crate::task::Task;
+
+type Receiver = tokio::sync::mpsc::Receiver<tedge_api::messages::Message>;
+type Sender = tokio::sync::mpsc::Sender<tedge_api::messages::Message>;
```

Comment:
As I undersand these are the channels for Messaging from Plugin to Core and vice versa. Right?

I see that's one technical reason that requires to keep all plugins using those channels in one process. Is that right? I would really pefer to get rid of that restriction.

On option would be to use some similar Rust API that works on IPC level (if there is one). Another option would be to implement some message bridging from those Rust channels to some IPC (e.g. Unix Domain Socket). Finally I feel there is a way around.

Is there any other technical need to keep plugins in one process, that I currently not see?

Reply:
> As I undersand these are the channels for Messaging from Plugin to Core and vice versa. Right?

Not entirely, a plugin receives a message via the `Plugin::handle_message()` function. The channels are used for sending a message from a plugin to the core, but not the other way round. Also, the `tedge_core` crate uses channels internally while orchestrating the plugins.

> I see that's one technical reason that requires to keep all plugins using those channels in one process. Is that right?

It is right that the channels are used inside one process, but the restriction does not exist because one can very easily implement a Plugin that translates between process-internal message to, for example, some IPC mechanism (as you suggested above, of course)!

Reply:
> Is there any other technical need to keep plugins in one process, that I currently not see?

Not necessarily a technical one, but we get some very nice benefits in regards to maintainability and orchestration out of this!

* Less overhead while communicating (copying pointers basically)
* Smaller overall binary size (one binary vs many)
* Tight coupling to ensure that all plugins (rather: components) speak the same "language", while still having the option to have loose coupling via translation of messages to external processes (via IPC as discussed above for example)
* Easier orchestration, as the whole lifecycle of a plugin is well-defined
  * Setup of components of the software are well-defined
  * Shutdown of components of the software are well-defined
* Easier installation, as one only has to deploy one binary and a configuration file, not many.

Reply:
> > Is there any other technical need to keep plugins in one process, that I currently not see?
> 
> Not necessarily a technical one, but we get some very nice benefits in regards to maintainability and orchestration out of this!
> 
> * Less overhead while communicating (copying pointers basically)
> * ... maybe more I did not think about right now 😆

For me the main benefit is that this imposes nothing to the users. And if required for some use case, one can always add a connector plugin to bridge a thin-edge executable with an IPC protocol. By contract, if the thin-edge core embarks per design an IPC+serialization protocol then this protocol layer *must* be available. Again a bridge can be added, but then the two protocols must be available.